### PR TITLE
feat(scanners): port semantic scanner package to sdk/python/acf/scanners

### PR DESCRIPTION
**Pull request by Pranjal0410**
First step toward the Python-to-Go scanner integration. Brings the semantic scanner package into the current SDK layout so the next PR can wire its signals through to the sidecar.

## What this PR does

Ports the semantic scanner package I wrote back in March into sdk/python/acf/scanners. The code is intact — same pluggable backend (sentence-transformer for prod, tf-idf + SVD for CI), same pre-computed attack embeddings, same signal-producer output. Just sitting in the new layout now with its tests.

## What this PR doesn't do

No integration with the sidecar yet. No changes to the Firewall class. No changes to the wire protocol. The scanner is in the package but nothing calls it. That wiring is the next PR.

## Files

- semantic_scanner.py - main entry point, runs after lexical scan when input is still untrusted
- backends.py - pluggable embedding backend
- attack_library.py - curated attack patterns across 6 categories
- models.py - Pydantic models for ScanInput, SemanticHit, SemanticScannerOutput
- __init__.py - public API

## Dependencies

Scanner deps (pydantic, numpy, sentence-transformers, scikit-learn) added as an optional [scanners] extra in pyproject.toml. Base SDK stays zero-deps as before. To use the scanner: pip install -e .[scanners].

## Tests

22 passing in sdk/python/tests/test_semantic_scanner.py. Covers known attacks, benign inputs, block thresholds, configuration, edge cases, latency, and output contract. Full suite still 78/78 passing.

## Next PR

SDK to sidecar wiring - Firewall pre-populates signals in RiskContext before sending, so OPA sees the semantic hits alongside the lexical signals.

### CODE REVIEW

**Comment by kavishkafer** [comment]
Hey @Pranjal0410  Can you double check is this a logic bug : how `block_threshold` is evaluated.
`hits` is built by filtering `similarities` against `default_threshold` 
`category_thresholds` first (line 556). The `block_threshold` check then
only iterates over that already-filtered list.   @tharindupr

File: sdk/python/acf/scanners/semantic_scanner.py [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+            )
+            for idx, sim in hits
+        ]
+
+        max_similarity = float(np.max(similarities)) if len(similarities) > 0 else 0.0
+        risk_score = round(max(max_similarity, 0.0), 4)
+
+        action = ScanAction.PROCEED
+        reason = None
+        if any(sim >= self._config.block_threshold for _, sim in hits):
```
Comment (kavishkafer) [inline sdk/python/acf/scanners/semantic_scanner.py]: block_threshold check runs against the already-filtered `hits` list,
not against all similarities.
`hits` is built at  by filtering `similarities` using
`default_threshold` / `category_thresholds`. So any pattern whose similarity falls between `block_threshold` and `default_threshold` never enters `hits` and therefore never triggers a block  even though it should.
For example:
 `default_threshold = 0.85`, `block_threshold = 0.80`
 A pattern scores 0.82  excluded from `hits` 
Block check  never sees it  `PROCEED` is returned instead of SHORT_CIRCUIT_BLOCK

**Comment by Pranjal0410** [comment]
Thanks @kavishkafer, you were right. block check now runs against all similarities directly, and the blocking pattern shows up in semantic_hits even if it's below the default threshold. added a regression test that would've failed before the fix

File: sdk/python/acf/scanners/backends.py [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+    def __init__(self, n_components: int = 128) -> None:
+        from sklearn.feature_extraction.text import TfidfVectorizer
+        from sklearn.decomposition import TruncatedSVD
+
+        self._vectorizer = TfidfVectorizer(
+            max_features=5000,
+            ngram_range=(1, 3),
+            sublinear_tf=True,
+        )
+        self._svd = TruncatedSVD(n_components=n_components, random_state=42)
```
Comment (kavishkafer) [inline sdk/python/acf/scanners/backends.py:97]: When we call fit(self, corpus) with the standard attack library, the corpus only contains 49 documents. When tfidf_matrix executes, it will throw a ValueError at runtime
To prevent this crash when the corpus size is smaller than the requested components, we should dynamically size the SVD components during fit() @tharindupr

Comment (kavishkafer) [reply sdk/python/acf/scanners/backends.py:97]: Others things are  LGTM!

**Comment by Pranjal0410** [comment]
verified and fixed @kavishkafer. small corpora do crash (n_components must be < min(n_samples, n_features)) -clamping dynamically in fit() now. added a 3-doc regression test for it. 24 passing.

**Comment by kavishkafer** [comment]
LGTM!

**Comment by tharindupr** [comment]
I will merge this for now. We need a config parameter to turn this feature on and off. After wiring this into the scanner, we should measure how this feature improved the detection rate.
